raft.go

```diff
@@ -308,7 +315,54 @@ func (r *Raft) runCandidate() {
 		case rpc := <-r.rpcCh:
 			r.mainThreadSaturation.working()
 			r.processRPC(rpc)
+		case vote := <-prevoteCh:
+			// This a pre-vote case but could lead to winning an election, in the case that majority of other nodes
+			// don't support pre-vote or have pre-vote deactivated.
+			r.mainThreadSaturation.working()
+			r.logger.Debug("got a prevote!!", "from", vote.voterID, "term", vote.Term, "tally", grantedVotes)
+			// Check if the term is greater than ours, bail
+			if vote.Term > r.getCurrentTerm() && !vote.PreVote {
+				r.logger.Debug("newer term discovered, fallback to follower", "term", vote.Term)
+				r.setState(Follower)
+				r.setCurrentTerm(vote.Term)
+				return
+			}
+			if vote.Term > term && vote.PreVote {
+				r.logger.Debug("newer term discovered on pre-vote, fallback to follower", "term", vote.Term)
+				r.setState(Follower)
+				r.setCurrentTerm(vote.Term)
+				return
+			}
+
+			// Check if the vote is granted
+			if vote.Granted {
+				if !vote.PreVote {
+					grantedVotes++
+					r.logger.Debug("vote granted", "from", vote.voterID, "term", vote.Term, "tally", grantedVotes)
+				} else {
+					preVoteGrantedVotes++
+					r.logger.Debug("prevote granted", "from", vote.voterID, "term", vote.Term, "tally", preVoteGrantedVotes)
+				}
+			}
+
+			// Check if we've become the leader
+			// If we won the election because the majority of nodes don't support pre-vote (or pre-vote is deactivated)
+			// set our state to leader and our term to the pre-vote term.
+			if grantedVotes >= votesNeeded {
+				r.logger.Info("election won", "term", vote.Term, "tally", grantedVotes)
+				r.setState(Leader)
+				r.setLeader(r.localAddr, r.localID)
 
+				r.setCurrentTerm(term)
+				return
+			}
+			// Check if we've won the pre-vote and proceed to election if so
+			if preVoteGrantedVotes >= votesNeeded {
+				r.logger.Info("pre election won", "term", vote.Term, "tally", preVoteGrantedVotes)
+				preVoteGrantedVotes = 0
+				grantedVotes = 0
+				voteCh = r.electSelf(false)
```

Comment:
Do we also need to reset the electionTimer? I'm not sure. If we don't it means that we have to get PreVote _and_ Vote rounds within one election timeout. If we are close to timing out when we get PreVotes through then it seems wasteful to start a real election that's really likely to fail due to timing out before responses are gathered and then restart prevoting all over again!

Reply:
yes that's probably needed and explain some of the timing issue I was having in tests.